### PR TITLE
[processor/resourcedetection] Add k8s cluster name detection in EKS environment

### LINKED ISSUE #26794: [processor/resourcedetection]: Add support for k8s.cluster.name where possible
### Component(s)

processor/k8sattributes

### Is your feature request related to a problem? Please describe.

When using the `k8sattributes` processor, I shouldn't need to declare any attributes manually. I still need to manually specify `k8s.cluster.name` as the processor doesn't support it yet.

There is no standard API to get the cluster name, but GKE and EKS make it easy to detect. We should support this in the processor and provide `k8s.cluster.name` where possible.

### Describe the solution you'd like

For GKE, See: https://github.com/open-telemetry/opentelemetry-js-contrib/blob/deb9aa441dc7d2b0fd5ec11b41c934a1e93134fd/detectors/node/opentelemetry-resource-detector-gcp/src/detectors/GcpDetector.ts#L81 (it gets it from `curl http://metadata/computeMetadata/v1/instance/attributes/cluster-name -H "Metadata-Flavor: Google"`)

For EKS see: https://github.com/open-telemetry/opentelemetry-js-contrib/blob/deb9aa441dc7d2b0fd5ec11b41c934a1e93134fd/detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEksDetector.ts#L86 (it gets it from the cert name on `kubernetes.default.svc`, iiuc)

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by crobert-1**
**Description:** <Describe what has changed.>

This enhancement detects the k8s cluster name in EKS. The solution uses EC2 instance tags to determine the cluster name, which means it will only work on EC2 (as noted in documentation updates).

Note: This is the second attempt at adding this functionality, superseding PR #27408. My first attempt was using `v2` of the [AWS GO SDK](https://github.com/aws/aws-sdk-go-v2), but the `service/ec2` module was causing lint to hit OOM issues and successful lint runs were taking over half an hour to complete. This uses `v1` of the module instead in an attempt to resolve linting issues.

**Link to tracking Issue:** <Issue number if applicable>
Resolves #26794

**Testing:** <Describe what testing was performed and which tests were added.>
Manual testing in an EKS cluster when running on EC2 instances. Added unit tests to cover basic functionality.

**Documentation:** <Describe the documentation added.>
Updated README and relevant metadata

### CODE REVIEW

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 1]
```diff
@@ -74,7 +86,12 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
 	d.rb.SetCloudProvider(conventions.AttributeCloudProviderAWS)
 	d.rb.SetCloudPlatform(conventions.AttributeCloudPlatformAWSEKS)
 
-	return d.rb.Emit(), conventions.SchemaURL, nil
+	// The error is unhandled because we want to return successfully detected resources
+	// regardless of an error. The caller will properly handle any error hit while getting
+	// the cluster name.
+	clusterName, err := d.utils.getClusterName(ctx)
+	d.rb.SetK8sClusterName(clusterName)
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:93]: can you please wrap it in `if rb.config.K8sClusterName.Enabled {` so user don't run into errors issues if the atribute isn't needed

File: processor/resourcedetectionprocessor/internal/aws/eks/detector_test.go [hunk 2]
```diff
@@ -16,6 +17,11 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/resourcedetectionprocessor/internal/aws/eks/internal/metadata"
 )
 
+const (
+	clusterNameKey = "k8s.cluster.name"
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector_test.go:21]: `conventions.AttributeK8SClusterName` can be used instead

File: processor/resourcedetectionprocessor/factory.go [hunk 3]
```diff
@@ -97,7 +101,7 @@ func createDefaultConfig() component.Config {
 
 func defaultHTTPClientSettings() confighttp.HTTPClientSettings {
 	httpClientSettings := confighttp.NewDefaultHTTPClientSettings()
-	httpClientSettings.Timeout = 5 * time.Second
+	httpClientSettings.Timeout = defaultTimeout
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/factory.go:104]: please avoid unrelated changes

File: processor/resourcedetectionprocessor/README.md [hunk 4]
```diff
@@ -334,21 +338,21 @@ processors:
...
+[Cloud semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/cloud.md)
 
 * `cloud.provider` (`"aws"`)
 * `cloud.platform` (`"aws_lambda"`)
 * `cloud.region` (`$AWS_REGION`)
 
-[Function as a Service semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/faas.md)
-and [AWS Lambda semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-lambda.md#resource-detector)
+[Function as a Service semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/faas.md)
+and [AWS Lambda semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/faas/aws-lambda.md)
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/README.md:348]: These fixes would be nice as a separate PR

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 5]
```diff
@@ -114,3 +131,49 @@ func (e eksDetectorUtils) getConfigMap(ctx context.Context, namespace string, na
 	}
 	return cm.Data, nil
 }
+
+func (e eksDetectorUtils) getClusterName(ctx context.Context) (string, error) {
+	sess := session.Must(session.NewSession())
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:136]: Can we avoid panic and return an error instead?

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 1 above)
Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:93]: Good call, will do.

Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:93]: Actually, it looks like we don't have access to this since `config` is a private member of the `ResourceBuilder` struct in the `metadata` package. Is it worth re-architecting some of this to get access, or should we leave it as is?

File: processor/resourcedetectionprocessor/README.md (hunk 4 above)
Comment (crobert-1) [reply processor/resourcedetectionprocessor/README.md:348]: Moved to #28669.

**Comment by crobert-1** [comment]
`check-links` failure will be fixed by #28669

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 1 above)
Comment (dmitryax) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:93]: It should be easy to pass. Just add a new field to `detector` struct and set `cfg.ResourceAttributes` to it in `NewDetector` func

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
@jinja2 can you please take a look as well?

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 6]
```diff
@@ -114,3 +137,53 @@ func (e eksDetectorUtils) getConfigMap(ctx context.Context, namespace string, na
...
+		return clusterName, fmt.Errorf("Failed to detect EKS cluster name. No tag for cluster name found on EC2 instance")
+	}
+	return clusterName, nil
+}
+
+func (e eksDetectorUtils) getClusterNameTagFromReservations(reservations []*ec2.Reservation) string {
+	for _, reservation := range reservations {
+		for _, instance := range reservation.Instances {
+			for _, tag := range instance.Tags {
+				if *tag.Key == clusterNameAwsEksTag || *tag.Key == clusterNameEksTag {
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:179]: Is the key guaranteed to be not `nil`?

Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:179]: I just checked, there's no enforcement in place to guarantee it's non-nil. I'll add a check here. 👍

File: processor/resourcedetectionprocessor/README.md [hunk 7]
```diff
@@ -318,14 +318,18 @@ processors:
...
+If you see an error with the message `context deadline exceeded`, please increase the timeout setting in your config.
 
 Example:
 
 ```yaml
 processors:
   resourcedetection/eks:
     detectors: [env, eks]
-    timeout: 2s
+    timeout: 15s
```
Comment (jinja2) [inline processor/resourcedetectionprocessor/README.md:332]: Is this the avg duration you noticed in your tests? I assume users will be mindful of this additional delay and update readiness check timeouts when the healthcheck extension + k8s probe are enabled.

Comment (crobert-1) [reply processor/resourcedetectionprocessor/README.md:332]: I was generally testing around `15s`, I think around `11s` was the lowest timeout that usually worked before timing out.

Do you think a comment should be added here regarding using this with the healthcheck extension?

Comment (jinja2) [reply processor/resourcedetectionprocessor/README.md:332]: Just checked the liveness probe values set by default in the helm-chart, we should be good if 15s is the worst observed (probe is run 3 times with 10sec interval before restarting). I was hoping the [healthcheck extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/healthcheckextension) had some documentation about things to consider for setting intervals/threshold for health probes, but no dice.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 8]
```diff
@@ -74,7 +88,16 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
...
+	// The error is unhandled because we want to return successfully detected resources
+	// regardless of an error. The caller will properly handle any error hit while getting
+	// the cluster name.
+	if d.ra.K8sClusterName.Enabled {
+		var clusterName string
+		clusterName, err = d.utils.getClusterName(ctx)
+		d.rb.SetK8sClusterName(clusterName)
+	}
+
+	return d.rb.Emit(), conventions.SchemaURL, err
```
Comment (bryan-aguilar) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:100]: Who is the caller in this case and how is it handled? Is the caller built in such a way where it treats this as a partial error and doesn't immediately exit on error?

Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:100]: Good point, caller is here:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b716a4d89acd497cb1ed044b941c49a7897b72bb/processor/resourcedetectionprocessor/internal/resourcedetection.go#L129-L134

The error will result in the correctly detected resources not being returned. I'll add a warning log message instead of returning an error here, I believe that's the best way to balance this.

Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go]: Tested with update, here's the logging output:
```
2023-11-09T23:03:04.703Z	info	internal/resourcedetection.go:125	began detecting resource information	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs"}
2023-11-09T23:03:07.846Z	warn	eks/detector.go:97	Unable to get EKS cluster name	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs", "error": "EC2MetadataRequestError: failed to get EC2 instance identity document\ncaused by: RequestCanceled: request context canceled\ncaused by: context deadline exceeded"}
2023-11-09T23:03:07.846Z	info	internal/resourcedetection.go:139	detected resource information	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs", "resource": {"cloud.platform":"aws_eks","cloud.provider":"aws"}}
```
We are able to receive successfully detected resources, with the log message for the `k8s.cluster.name` failure.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 9]
```diff
@@ -74,7 +88,20 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
 	d.rb.SetCloudProvider(conventions.AttributeCloudProviderAWS)
 	d.rb.SetCloudPlatform(conventions.AttributeCloudPlatformAWSEKS)
 
-	return d.rb.Emit(), conventions.SchemaURL, nil
+	if d.ra.K8sClusterName.Enabled {
```
Comment (bryan-aguilar) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:91]: Do we need to perform this check ourselves? It looks like the generated code does that for you inside `SetK8sClusterName`. I could see the benefit of doing the check, that way we only perform the query if it's enabled. It could confuse users if the see an error for cluster name but they don't have it enabled.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 10]
```diff
@@ -74,7 +88,20 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
...
+		var clusterName string
+		clusterName, err = d.utils.getClusterName(ctx)
+		if err != nil {
+			// We clear out the error here so that the correctly detected resources
+			// can still be used by the processor.
+			d.logger.Warn("Unable to get EKS cluster name", zap.Error(err))
+			err = nil
+		} else {
+			d.rb.SetK8sClusterName(clusterName)
+		}
```
Comment (bryan-aguilar) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:101]: Is this going to log each time a datapoint is processed or only once on startup? If it's each time then we should consider making this debug and not warn level. 

Also, manually setting the error here feels like an anti pattern. If the error is only used for a logging statement, and never returned to the caller, would it make more sense to not return an error at all from `getClusterName`? A logging statement could just be added inside that function.

Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:101]: This is only logged once at collector startup.

Example logs: (collector is running with filelog receiver and debug exporter)
```
$ kubectl -n crobert-test logs otel-collector-7985f54766-dtdbf
2023-11-10T16:55:22.020Z	info	service@v0.88.1-0.20231026220224-6405e152a2d9/telemetry.go:84	Setting up own telemetry...
2023-11-10T16:55:22.020Z	info	service@v0.88.1-0.20231026220224-6405e152a2d9/telemetry.go:201	Serving Prometheus metrics	{"address": ":8888", "level": "Basic"}
2023-11-10T16:55:22.020Z	info	exporter@v0.88.1-0.20231026220224-6405e152a2d9/exporter.go:275	Development component. May change in the future.	{"kind": "exporter", "data_type": "logs", "name": "debug"}
2023-11-10T16:55:22.110Z	info	service@v0.88.1-0.20231026220224-6405e152a2d9/service.go:143	Starting otelcontribcol...	{"Version": "0.88.0-dev", "NumCPU": 16}
2023-11-10T16:55:22.110Z	info	extensions/extensions.go:33	Starting extensions...
2023-11-10T16:55:22.110Z	info	internal/resourcedetection.go:125	began detecting resource information	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs"}
2023-11-10T16:55:25.310Z	warn	eks/detector.go:97	Unable to get EKS cluster name	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs", "error": "EC2MetadataRequestError: failed to get EC2 instance identity document\ncaused by: RequestCanceled: request context canceled\ncaused by: context deadline exceeded"}
2023-11-10T16:55:25.310Z	info	internal/resourcedetection.go:139	detected resource information	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs", "resource": {"cloud.platform":"aws_eks","cloud.provider":"aws"}}
2023-11-10T16:55:25.310Z	info	adapter/receiver.go:45	Starting stanza receiver	{"kind": "receiver", "name": "filelog", "data_type": "logs"}
2023-11-10T16:55:25.310Z	info	service@v0.88.1-0.20231026220224-6405e152a2d9/service.go:169	Everything is ready. Begin running and processing data.
2023-11-10T16:55:25.511Z	info	fileconsumer/file.go:251	Started watching file	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer", "path": "/var/log/pods/analytics-tests-rv_quantizer-alpha-0_551d1187-3e40-4179-a5fb-3b752da44ad6/quantizer/0.log"}
2023-11-10T16:55:28.811Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 5}
2023-11-10T16:55:32.611Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 2}
2023-11-10T16:55:41.810Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 3}
2023-11-10T16:55:44.610Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 2}
2023-11-10T16:55:50.211Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 2}
2023-11-10T16:56:00.211Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 6}
2023-11-10T16:56:01.210Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 4}
2023-11-10T16:56:02.811Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 10}
2023-11-10T16:56:04.011Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 6}
2023-11-10T16:56:05.011Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 4}
2023-11-10T16:56:06.713Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 100}
2023-11-10T16:56:06.811Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 14}
2023-11-10T16:56:08.611Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 1}
2023-11-10T16:56:08.810Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 3}
2023-11-10T16:56:11.411Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 3, "log records": 7}
2023-11-10T16:56:11.811Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 6}
2023-11-10T16:56:15.211Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 7}
2023-11-10T16:56:16.211Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 4}
```
I agree the manual setting of `err = nil` should be improved. I think it would be fair to log each error in `getClusterName` at the warn level, and simply return an empty string. No harm is done by setting the k8s.cluster.name to be empty, so I that sounds like a fair solution to me.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 9 above)
Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:91]: Yes, you're hitting on the reasons why we added this check explicitly. The logic to get the k8s cluster name can take 10+ seconds, and will often hit errors when running with the default timeout (or always hit errors if user doesn't have proper permission). Adding this means the user won't have the long delay and unrelated errors if the cluster name is disabled.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 11]
```diff
@@ -74,7 +88,12 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
 	d.rb.SetCloudProvider(conventions.AttributeCloudProviderAWS)
 	d.rb.SetCloudPlatform(conventions.AttributeCloudPlatformAWSEKS)
 
-	return d.rb.Emit(), conventions.SchemaURL, nil
+	if d.ra.K8sClusterName.Enabled {
+		clusterName := d.utils.getClusterName(ctx, d.logger)
+		d.rb.SetK8sClusterName(clusterName)
+	}
+
+	return d.rb.Emit(), conventions.SchemaURL, err
```
Comment (bryan-aguilar) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:96]:
```suggestion
	return d.rb.Emit(), conventions.SchemaURL, nil
```

The only reason I am suggesting this is that the `isEKS()` doesn't actually handle the error directly. I'm not sure if it's possible but there could be a scenario where `isEKS == true` and `err != nil`. In this case you would be returning an error here and the resources would be dropped by caller. 

I think we could return nil here and/or check for non nil error above.

Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:96]: Fair point. There currently isn't any scenario where `err` would be non-nil, and `isEKS` would be true, but I could see the confusion.

**Review by bryan-aguilar — APPROVED** [review]

**Comment by crobert-1** [comment]
Lint failure unrelated (fixed by #29101), signalfx exporter test failure is also unrelated (added frequency to #29099)

**Review by jinja2 — APPROVED** [review]
